src/frontend/specification/ParameterTable.jl

```diff
@@ -124,128 +153,182 @@ end
-    from = partable.columns[:from][is_regression]
+    # regression edges (excluding intercept)
+    edges = [
+        (from, to) for (rel, from, to) in zip(
+            partable.columns[:relation],
+            partable.columns[:from],
+            partable.columns[:to],
+        ) if (rel == :→) && (from != Symbol("1"))
+    ]
+    sort!(edges, by = last) # sort edges by target
 
-    sorted_variables = Vector{Symbol}()
+    sorted_vars = Vector{Symbol}()
 
-    sorted = false
-    while !sorted
+    while !isempty(vars)
         acyclic = false
 
-        for (i, variable) in enumerate(variables)
-            if !(variable ∈ to)
-                push!(sorted_variables, variable)
-                deleteat!(variables, i)
-                delete_edges = from .!= variable
-                to = to[delete_edges]
-                from = from[delete_edges]
+        for (i, var) in enumerate(vars)
+            # check if var has any incoming edge
+            eix = searchsortedfirst(edges, (var, var), by = last)
+            if !(eix <= length(edges) && last(edges[eix]) == var)
+                # var is source, no edges to it
+                push!(sorted_vars, var)
+                deleteat!(vars, i)
+                # remove var outgoing edges
+                filter!(e -> e[1] != var, edges)
                 acyclic = true
+                break
             end
         end
 
-        if !acyclic
+        # if acyclic is false, all vars have incoming edge
+        acyclic ||
             throw(CyclicModelError("your model is cyclic and therefore can not be ordered"))
-        end
-        acyclic = false
-
-        if length(variables) == 0
-            sorted = true
-        end
     end
 
-    push!(partable.variables, :sorted_vars => sorted_variables)
+    copyto!(resize!(partable.sorted_vars, length(sorted_vars)), sorted_vars)
+    @assert length(partable.sorted_vars) == length(partable.observed_vars) + length(partable.latent_vars)
 
     return partable
 end
 
-function sort(partable::ParameterTable)
-    new_partable = deepcopy(partable)
-    sort!(new_partable)
-    return new_partable
-end
+"""
+    sort_vars(partable::ParameterTable)
+    sort_vars(partables::EnsembleParameterTable)
 
-# add a row --------------------------------------------------------------------------------
+Sort variables in `partable` so that all independent variables are
+before the dependent variables, and return the copy of `partable`.
 
-import Base.push!
+See [sort_vars!](@ref) for in-place version.
+"""
+sort_vars(partable::ParameterTable) = sort_vars!(deepcopy(partable))
+
+# add a row --------------------------------------------------------------------------------
 
-function push!(partable::ParameterTable, d::AbstractDict)
-    for key in keys(d)
-        push!(partable.columns[key], d[key])
+function Base.push!(partable::ParameterTable, d::Union{AbstractDict{Symbol}, NamedTuple})
+    for (key, val) in pairs(d)
```

Comment:
I think we should prevent people from adding partable rows that have more / less keys than columns, as this will lead to some quite weird behavior...
```suggestion
function Base.push!(partable::ParameterTable, d::Union{AbstractDict{Symbol}, NamedTuple})
    issetequal(keys(a), keys(b)) || throw(ArgumentError("The new row needs to have the same keys as the columns of the parameter table."))
    for (key, val) in pairs(d)
```